pkg/adapterManager/manager.go

```diff
@@ -149,10 +159,27 @@ func (m *Manager) Quota(ctx context.Context, requestBag *attribute.MutableBag, r
 	return qmr, o
 }
 
+// Preprocess dispatches to the set of aspects that must run before any other
+// configured aspects.
+func (m *Manager) Preprocess(ctx context.Context, requestBag, responseBag *attribute.MutableBag) rpc.Status {
+	return m.dispatch(ctx, requestBag, responseBag, m.preprocessKindSet,
+		func(executor aspect.Executor, eval expr.Evaluator) rpc.Status {
+			ppw := executor.(aspect.PreprocessExecutor)
+			result, rpcStatus := ppw.Execute(requestBag, eval)
+			if status.IsOK(rpcStatus) {
+				if err := responseBag.Merge(result.Attrs); err != nil {
+					// TODO: better error messages that push internal details into debuginfo messages
+					rpcStatus = status.WithInternal("The results from the request preprocessing could not be merged.")
+				}
+			}
+			return rpcStatus
+		})
+}
+
 type invokeExecutorFunc func(executor aspect.Executor, evaluator expr.Evaluator) rpc.Status
 
 // dispatch resolves config and invokes the specific set of aspects necessary to service the current request
-func (m *Manager) dispatch(ctx context.Context, requestBag *attribute.MutableBag, responseBag *attribute.MutableBag,
+func (m *Manager) dispatch(ctx context.Context, requestBag, responseBag *attribute.MutableBag,
```

Comment:
Dispatch ends up calling `config.Resolve()`
When the call chain is `Preprocess --> dispatch` it is not safe to do so `config.Resolve()`


I think we need to change the structure. 
`Preprocess` should only use adapters (because they do not have dependencies on attributes)

`Manager.dispatch` can implement request lifecycle so it will look like this
```
func dispatch () {
	cfg, _ := m.cfg.Load().(config.Resolver)

        m.Preprocess(cfg)    //  <-- do not call resolve on config
                                         // we can only safely access all descriptors and other globals.
                                         // Add new type of global called Preprocessor 

	cfgs, err := cfg.Resolve(requestBag, kindSet)
}

func Preprocess (cfg) {
      adapters := cfg.GetPreProcessAdapters()

      foreach adapter in adapters:
         adapter.preprocess()
}
```